keps/sig-scheduling/2249-pod-affinity-namespace-selector/README.md

```diff
@@ -0,0 +1,662 @@
+
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Allow users to dynamically select the set of namespaces considered when using pod
+  affinity/anti-affinity
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+<!--
+### Non-Goals
+
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+I am running a SaaS service, the workload for each customer is placed in a separate
+namespace. The workloads requires 1:1 pod to node placement. I want to use pod 
+anti-affinity across all customers namespaces to achieve that.
+
+<!--
+
+### Notes/Constraints/Caveats (Optional)
+
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Performance
+Using namespace selector will make it easier for users to specify affinity 
+constraints across a large number of namespaces. The initial implementation 
+of pod affinity/anti-affinity suffered from performance challenges, however 
+over the releases 1.17 - 1.20 we significantly improved that. 
+
+We currently have integration test and clusterloader benchmarks that evaluate the extreme
+cases of all pods having affinity/anti-affinity constraints to/against each other.
+Those benchmarks show that the scheduler is able to achieve maximum throughput 
+(i.e., the api-server qps limit).
+
+#### Security/Abuse
+see previous discussion [here](https://github.com/kubernetes/kubernetes/issues/68827#issuecomment-468470598)
+NamespaceSelector will allow users to select all namespaces. This may cause a security 
+concern: a pod with anti-affinity constraint can block pods from all other namespaces 
+from getting scheduled in a failure domain.
```

Comment:
the limited operations/validation below helps mitigate disrupting all namespaces... I can't tell if this enhancement makes it easier to disrupt specific namespaces, or is that already possible via that existing anti-affinity namespace field.

Is there opt-in required on both sides for pod scheduling to be disrupted, or can a user in namespace `foo` currently prevent regular `kube-system` pods from being scheduled by setting up an anti-affinity rule like:

```yaml
podAntiAffinity:
  requiredDuringSchedulingIgnoredDuringExecution:
  - labelSelector: {}
    namespaces: ["kube-system"]
```

Reply:
Are we happy with that? That seems like a mistake in the existing feature to me... do we want to continue to allow that with this new capability or try to address that possibility for cross-namespace interference?

Reply:
> Are we happy with that? That seems like a mistake in the existing feature to me... 

Are we able to fix this now? Wouldn't restricting `labelSelector` now be a backward incompatible change?

> do we want to continue to allow that with this new capability or try to address that possibility for cross-namespace interference?

You mean limit labelSelector only if namespaceSelector is set? I am fine with that, but wouldn't that surprise users that it is allowed in some cases and not in others?

Reply:
> No opt-in on both sides, an existing pod can already prevent other pods from being scheduled on the node, the example you gave would prevent kube-system pods from being scheduled on the node where the pod with that term is scheduled.

On the flip side, it prevents the incoming Pod running on a Node where there is already a kube-system Pod scheduled.

Reply:
> On the flip side, it prevents the incoming Pod running on a Node where there is already a kube-system Pod scheduled.

Yeah, but that is not really useful. Separating system from workload pods should be done using taint/tolerations and perhaps node affinity.

btw, since system pods are usually created with the highest priority, the existing pod will likely get preempted if it is preventing the system pod from getting scheduled via anti-affinity.

Reply:
> > Are we happy with that? That seems like a mistake in the existing feature to me...
> 
> Are we able to fix this now? Wouldn't restricting `labelSelector` now be a backward incompatible change?

The current proposal doubles down on letting one namespace unilaterally prevent scheduling of pods from another namespace. Anyone who currently didn't want to allow that in their cluster would have put some sort of custom webhook admission policy check in place to prevent affinity rules using the `namespaces` field. This proposal adds a new field that exposes the same cross-namespace capability.

Put another way, a cluster someone had locked down to prevent cross-namespace interference would be vulnerable again after this was implemented unless they notice and make changes to their custom policy enforcement.

> > do we want to continue to allow that with this new capability or try to address that possibility for cross-namespace interference?
> 
> You mean limit labelSelector only if namespaceSelector is set? I am fine with that, but wouldn't that surprise users that it is allowed in some cases and not in others?

No, I wasn't suggesting changing validation between the fields. I was asking if use of this cross-namespace capability should be opt-in for the cluster admin, rather than opt-out.

Reply:
Most clusters would want this enabled, so I don't think we want this feature to be opt-in. The feature will be disabled by default in alpha, so that will give people a full release to make the change to their custom policies. So I suggest we keep it opt-in and stress the point regarding custom policies in the release notes, what do you think?

Reply:
Is there an alternative way to make an API opt-in beyond alpha? That is, when the feature gate is enabled by default.

Reply:
It's not exactly opt-in, but defining a quota scope for cross-namespace affinity/anti-affinity pods would let a cluster admin easily set up a default cluster wide quota to say "by default, cross-namespace affinity/antiaffinity pods are not allowed unless I explicitly grant quota for them in a namespace". That would be coherent with the approach to priorityclass pods

Reply:
Discussed offline. Practically adding a scope is like an on/off knob to control which namespaces can include pods that set namespaces/namespaceSelector fields (the actual hard limit is not going to be that important since a single pod can do the damage we are concerned about).

I will update the KEP to include proposing a new quota scope, so NamespaceSelector and "CrossNamespaceAffinity" will be done under one feature flag.

Reply:
I updated the KEP to include the new quota scope. PTAL.